File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/frontends/LanguageFrontend.java
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.frontends;

import de.fraunhofer.aisec.cpg.TranslationConfiguration;
import de.fraunhofer.aisec.cpg.graph.CodeAndLocationProvider;
import de.fraunhofer.aisec.cpg.graph.LanguageProvider;
import de.fraunhofer.aisec.cpg.graph.Node;
import de.fraunhofer.aisec.cpg.graph.declarations.TranslationUnitDeclaration;
import de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager;
import de.fraunhofer.aisec.cpg.sarif.PhysicalLocation;
import de.fraunhofer.aisec.cpg.sarif.Region;
import java.io.File;
import java.util.*;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class LanguageFrontend extends ProcessedListener
    implements CodeAndLocationProvider, LanguageProvider {

  // Allow non-Java frontends to access the logger (i.e. jep)
  public static final Logger log = LoggerFactory.getLogger(LanguageFrontend.class);

  protected Language<? extends LanguageFrontend> language;

  protected final TranslationConfiguration config;

  protected ScopeManager scopeManager;

  private final String namespaceDelimiter;
  protected TranslationUnitDeclaration currentTU = null;

  // Todo Moving this to scope manager and add listeners and processedMappings to specified scopes.

  public LanguageFrontend(
      @NotNull Language<? extends LanguageFrontend> language,
      @NotNull TranslationConfiguration config,
      ScopeManager scopeManager,
      String namespaceDelimiter) {
    this.language = language;
    this.config = config;
    this.namespaceDelimiter = namespaceDelimiter;
    this.scopeManager = scopeManager;
    this.scopeManager.setLang(this);
  }

  public List<TranslationUnitDeclaration> parseAll() throws TranslationException {
    ArrayList<TranslationUnitDeclaration> units = new ArrayList<>();
    for (String component : this.config.getSoftwareComponents().keySet()) {
      for (File sourceFile : this.config.getSoftwareComponents().get(component)) {
        units.add(parse(sourceFile));
      }
    }

    return units;
  }

  @NotNull
  public ScopeManager getScopeManager() {
    return scopeManager;
  }

  public void setScopeManager(@NotNull ScopeManager scopeManager) {
    this.scopeManager = scopeManager;
    this.scopeManager.setLang(this);
  }

  public TranslationUnitDeclaration getCurrentTU() {
    return currentTU;
  }

  public void setCurrentTU(TranslationUnitDeclaration currentTU) {
    this.currentTU = currentTU;
  }

  public abstract TranslationUnitDeclaration parse(File file) throws TranslationException;

  /**
   * Returns the raw code of the ast node, generic for java or c++ ast nodes.
   *
   * @param <T> the raw ast type
   * @param astNode the ast node
   * @return the source code
   */
  @Nullable
  public abstract <T> String getCodeFromRawNode(T astNode);

  /**
   * Returns the {@link Region} of the code with line and column, index starting at 1, generic for
   * java or c++ ast nodes.
   *
   * @param <T> the raw ast type
   * @param astNode the ast node
   * @return the location
   */
  @Nullable
  public abstract <T> PhysicalLocation getLocationFromRawNode(T astNode);

  public <N, S> void setCodeAndLocation(@NotNull N cpgNode, @Nullable S astNode) {
    if (cpgNode instanceof Node && astNode != null) {
      if (config.codeInNodes) {
        // only set code, if its not already set or empty
        String code = getCodeFromRawNode(astNode);

        if (code != null) {
          ((Node) cpgNode).setCode(code);
        } else {
          log.warn("Unexpected: No code for node {}", astNode);
        }
      }
      ((Node) cpgNode).setLocation(getLocationFromRawNode(astNode));
    }
  }

  /**
   * To prevent issues with different newline types and formatting.
   *
   * @param node - The newline type is extracted from the nodes code.
   * @return the String of the newline
   */
  public String getNewLineType(Node node) {
    List<String> nls = Arrays.asList("\n\r", "\r\n", "\n");
    for (String nl : nls) {
      if (node.toString().endsWith(nl)) {
        return nl;
      }
    }
    log.debug("Could not determine newline type. Assuming \\n. {}", node);
    return "\n";
  }

  /**
   * Returns the code represented by the subregion extracted from the parent node and its region.
   *
   * @param node - The parent node of the subregion
   * @param nodeRegion - region needs to be precomputed.
   * @param subRegion - precomputed subregion
   * @return the code of the subregion.
   */
  public String getCodeOfSubregion(
      de.fraunhofer.aisec.cpg.graph.Node node, Region nodeRegion, Region subRegion) {
    String code = node.getCode();
    if (code == null) {
      return "";
    }
    String nlType = getNewLineType(node);
    int start;
    int end;
    if (subRegion.getStartLine() == nodeRegion.getStartLine()) {
      start = subRegion.getStartColumn() - nodeRegion.getStartColumn();
    } else {
      start =
          StringUtils.ordinalIndexOf(
                  code, nlType, subRegion.getStartLine() - nodeRegion.getStartLine())
              + subRegion.getStartColumn();
    }
    if (subRegion.getEndLine() == nodeRegion.getStartLine()) {
      end = subRegion.getEndColumn() - nodeRegion.getStartColumn();
    } else {
      end =
          StringUtils.ordinalIndexOf(
                  code, nlType, subRegion.getEndLine() - nodeRegion.getStartLine())
              + subRegion.getEndColumn();
    }
    return code.substring(start, end);
  }

  /**
   * Merges two regions. The new region contains both and is the minimal region to do so.
   *
   * @param regionOne the first region
   * @param regionTwo the second region
   * @return the merged region
   */
  public Region mergeRegions(Region regionOne, Region regionTwo) {
    Region ret = new Region();
    if (regionOne.getStartLine() < regionTwo.getStartLine()
        || regionOne.getStartLine() == regionTwo.getStartLine()
            && regionOne.getStartColumn() < regionTwo.getStartColumn()) {
      ret.setStartLine(regionOne.getStartLine());
      ret.setStartColumn(regionOne.getStartColumn());
    } else {
      ret.setStartLine(regionTwo.getStartLine());
      ret.setStartColumn(regionTwo.getStartColumn());
    }
    if (regionOne.getEndLine() > regionTwo.getEndLine()
        || regionOne.getEndLine() == regionTwo.getEndLine()
            && regionOne.getEndColumn() > regionTwo.getEndColumn()) {
      ret.setEndLine(regionOne.getEndLine());
      ret.setEndColumn(regionOne.getStartColumn());
    } else {
      ret.setEndLine(regionTwo.getEndLine());
      ret.setEndColumn(regionTwo.getEndColumn());
    }
    return ret;
  }

  public void cleanup() {
    clearProcessed();
  }

  public String getNamespaceDelimiter() {
    return namespaceDelimiter;
  }

  public abstract <S, T> void setComment(S s, T ctx);

  public TranslationConfiguration getConfig() {
    return config;
  }

  @NotNull
  public Language<? extends LanguageFrontend> getLanguage() {
    return this.language;
  }
}

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/LanguageFrontend.kt
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.frontends

import de.fraunhofer.aisec.cpg.*
import de.fraunhofer.aisec.cpg.graph.*
import de.fraunhofer.aisec.cpg.graph.ast.declarations.TranslationUnitDeclaration
import de.fraunhofer.aisec.cpg.graph.scopes.Scope
import de.fraunhofer.aisec.cpg.graph.types.Type
import de.fraunhofer.aisec.cpg.sarif.PhysicalLocation
import de.fraunhofer.aisec.cpg.sarif.Region
import java.io.File
import java.util.*
import org.slf4j.LoggerFactory

/**
 * The main task of the language frontend is to translate the programming language-specific files to
 * the common CPG nodes. It further fills the scopeManager. The language frontend must not be used
 * after having processed the files, i.e., it won't be available in passes.
 *
 * More information can be found in the
 * [GitHub wiki page](https://github.com/Fraunhofer-AISEC/cpg/wiki/Language-Frontends).
 */
abstract class LanguageFrontend<AstNode, TypeNode>(
    /**
     * The translation context, which contains all necessary managers used in this frontend parsing
     * process. Note, that different contexts could be passed to frontends, e.g., in parallel
     * parsing to supply different managers to different frontends.
     */
    final override var ctx: TranslationContext,

    /** The language this frontend works for. */
    override val language: Language<out LanguageFrontend<AstNode, TypeNode>>,
) :
    ProcessedListener(),
    CodeAndLocationProvider<AstNode>,
    LanguageProvider,
    ContextProvider,
    ScopeProvider,
    NamespaceProvider,
    RawNodeTypeProvider<AstNode> {
    val scopeManager: ScopeManager = ctx.scopeManager
    val typeManager: TypeManager = ctx.typeManager
    val config: TranslationConfiguration = ctx.config

    var currentTU: TranslationUnitDeclaration? = null

    @Throws(TranslationException::class)
    fun parseAll(): List<TranslationUnitDeclaration> {
        val units = ArrayList<TranslationUnitDeclaration>()
        for (componentFiles in config.softwareComponents.values) {
            for (sourceFile in componentFiles) {
                units.add(parse(sourceFile))
            }
        }
        return units
    }

    @Throws(TranslationException::class) abstract fun parse(file: File): TranslationUnitDeclaration

    /**
     * This function returns a [TranslationResult], but rather than parsing source code, the
     * function [init] is used to build nodes in the Node Fluent DSL.
     */
    fun build(init: LanguageFrontend<*, *>.() -> TranslationResult): TranslationResult {
        return init(this)
    }

    /**
     * This function serves as an entry-point to type parsing in the language frontend. It needs to
     * return a [Type] object based on the ast type object used by the language frontend, e.g., the
     * parser.
     *
     * A language frontend will usually de-construct the ast type object, e.g., in case of pointer
     * or array types and then either recursively call this function or call other helper functions
     * similar to this one. Ideally, they should share the [typeOf] name, but have different method
     * signatures.
     */
    abstract fun typeOf(type: TypeNode): Type

    /**
     * Returns the raw code of the ast node, generic for java or c++ ast nodes.
     *
     * @param <T> the raw ast type
     * @param astNode the ast node
     * @return the source code </T>
     */
    abstract override fun codeOf(astNode: AstNode): String?

    /**
     * Returns the [Region] of the code with line and column, index starting at 1, generic for java
     * or c++ ast nodes.
     *
     * @param <T> the raw ast type
     * @param astNode the ast node
     * @return the location </T>
     */
    abstract override fun locationOf(astNode: AstNode): PhysicalLocation?

    open fun cleanup() {
        clearProcessed()
    }

    abstract fun setComment(node: Node, astNode: AstNode)

    companion object {
        // Allow non-Java frontends to access the logger (i.e. jep)
        val log = LoggerFactory.getLogger(LanguageFrontend::class.java)
    }

    override val scope: Scope?
        get() {
            return this.scopeManager.currentScope
        }

    override val namespace: Name?
        get() {
            return this.scopeManager.currentNamespace
        }
}
